applicatie/logic/aggregeren.py

```diff
@@ -0,0 +1,231 @@
+
+    # rekening baten aggregeren
+    data_agg, fouten = aggregeren_rekening(
+        data['baten'],
+        ['categorie', 2, 'categorie_baten'],
+        ['taakveld', 2, 'taakveld'],
+        codechecklijst
+    )
+    data_geaggregeerd.update({'baten': data_agg})
+    fouten_aggregeren.extend(fouten)
+
+    # rekening balans_baten aggregeren
+    data_agg, fouten = aggregeren_rekening(
+        data['balans_baten'],
+        ['categorie', 2, 'categorie_baten'],
+        ['balanscode', 2, 'balanscode'],
+        codechecklijst
+    )
+    data_geaggregeerd.update({'balans_baten': data_agg})
+    fouten_aggregeren.extend(fouten)
+
+    # rekening balans_standen aggregeren
+    data_agg, fouten = aggregeren_rekening(
+        data['balans_standen'],
+        ['standper', -1, 'standper'],  # aggregatieniveau -1 betekent niet aggregeren
+        ['balanscode', 2, 'balanscode'],
+        codechecklijst
+    )
+    data_geaggregeerd.update({'balans_standen': data_agg})
+    fouten_aggregeren.extend(fouten)
+
+    return data_geaggregeerd, fouten_aggregeren
+
+
+def aggregeren_rekening(data, dimensie_1, dimensie_2, codechecklijst):
+    """
+    data voor een rekening volledig aggregeren over twee dimensies
+    we specificeren hiervoor dimensie_1 en dimensie_2, dit zijn lists:
+    [ naam dimensie, aggregatieniveau, naam codelijst ]
+    """
+    foutenlijst = list()
+
+    # aggregeren over dimensie 1
+    vastedims = [dimensie_2[0], 'bedrag']   # we houden de tweede dimensie vast
+    aggdim = dimensie_1[0]
+    aggniv = dimensie_1[1]
+    clijst = codechecklijst.get(dimensie_1[2])
+    if aggniv > -1:
+        data, fouten = aggregeren_data(data, aggdim, vastedims, aggniv, clijst, False)
+        foutenlijst.extend(fouten)
+
+    # aggregeren over dimensie 2
+    vastedims = [dimensie_1[0], 'bedrag']   # we houden de eerste dimensie vast
+    aggdim = dimensie_2[0]
+    aggniv = dimensie_2[1]
+    clijst = codechecklijst.get(dimensie_2[2])
+    data, fouten = aggregeren_data(data, aggdim, vastedims, aggniv, clijst, True)
+    foutenlijst.extend(fouten)
+
+    data = data_opschonen(data, ['geteld', 'numgeteld'])
+
+    return data, foutenlijst
+
+
+def aggregeren_data(data, dimensie, vastedimensies, aggniveau, clijst, alleen_geteld):
+    """"aggregeren data: lijst van data records"""""
+    # Laatste wijziging 24 maart 2019 (getest in Jupyter notebook)
+
+    agg_data = list()
+    fouten = list()
+
+    if not clijst:
+        fouten.append("Geen codelijst beschikbaar voor {}.".format(dimensie))
+        return agg_data, fouten
+
+    if aggniveau < 0:
+        # stoppen indien aggregatieniveau negatief
+        fouten.append("Aggregatieniveau mag niet negatief zijn.")
+        return agg_data, fouten
+
+    cds = clijst[0::4]  # cds: lijst van codes
+    niv = clijst[2::4]  # niv: niveau van codes in de codelijst (0 = hoogste niveau)
+    ptc = clijst[3::4]  # ptc: lijst van parent codes behorende bij de codes
```

Comment:
Iets met sprongen van vier? Waarom zitten die sprongen er in?

Reply:
tjsa. dat gebeurt in de functie geef_codechecklijst()